Sources/GRPCHTTP2Core/GRPCStreamStateMachine.swift

```diff
@@ -0,0 +1,687 @@
+          }
+        }
+        
+        // All good
+      }
+
+      switch self.state {
+      case .clientIdleServerIdle(let state):
+        self.state = .clientOpenServerIdle(.init(
+          previousState: state,
+          compressionAlgorithm: metadata.encoding
+        ))
+        return .doNothing
+      case .clientOpenServerIdle, .clientOpenServerOpen, .clientOpenServerClosed:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Client shouldn't have sent metadata twice.")
+      case .clientClosedServerIdle, .clientClosedServerOpen, .clientClosedServerClosed:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Client can't have sent metadata if closed.")
+      }
+    }
+    
+    mutating func receive(message: ByteBuffer, endStream: Bool) throws {
+      switch self.state {
+      case .clientIdleServerIdle:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Can't have received a message if client is idle.")
+      case .clientOpenServerIdle(var state):
+        try state.deframer.process(buffer: message) { deframedMessage in
+          state.inboundMessageBuffer.append(deframedMessage)
+        }
+        
+        if endStream {
+          self.state = .clientClosedServerIdle(.init(previousState: state))
+        } else {
+          self.state = .clientOpenServerIdle(state)
+        }
+      case .clientOpenServerOpen(var state):
+        try state.deframer.process(buffer: message) { deframedMessage in
+          state.inboundMessageBuffer.append(deframedMessage)
+        }
+        
+        if endStream {
+          self.state = .clientClosedServerOpen(.init(previousState: state))
+        } else {
+          self.state = .clientOpenServerOpen(state)
+        }
+      case .clientOpenServerClosed:
+        // Client is not done sending request, but server has already closed.
+        // Ignore the rest of the request: do nothing.
+        ()
+      case .clientClosedServerIdle, .clientClosedServerOpen, .clientClosedServerClosed:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Client can't send a message if closed.")
+      }
+    }
+    
+    mutating func nextOutboundMessage() throws -> ByteBuffer? {
+      switch self.state {
+      case .clientIdleServerIdle, .clientOpenServerIdle, .clientClosedServerIdle:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Server is not open yet.")
+      case .clientOpenServerOpen(var state):
+        let response = try state.framer.next(compressor: state.compressor)
+        self.state = .clientOpenServerOpen(state)
+        return response
+      case .clientClosedServerOpen:
+        // No point in sending response if client is closed: do nothing.
+        return nil
+      case .clientOpenServerClosed, .clientClosedServerClosed:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Can't send response if server is closed.")
```

Comment:
There's no guarantee our outbound buffer has been drained yet so we shouldn't throw here